src/file/tempfile.cr

```diff
@@ -0,0 +1,138 @@
+class File
+  # Returns a fully-qualified path to a temporary file.
+  # The file is not actually created on the file system.
+  #
+  # ```
+  # File.tempname("foo", ".sock") # => "/tmp/foo20171206-1234-449386.sock"
+  # ```
+  #
+  # *prefix* and *suffix* are appended to the front and end of the file name, respectively.
+  # These values may contain directory separators.
+  #
+  # The path will be placed in *dir* which defaults to the standard temporary directory `Dir.tempdir`.
+  def self.tempname(prefix : String?, suffix : String?, *, dir : String = Dir.tempdir)
+    name = String.build do |io|
+      if prefix
+        io << prefix
+        io << '-'
+      end
+
+      io << Time.now.to_s("%Y%m%d")
+      io << '-'
+
+      {% unless flag?(:win32) %}
+        # TODO: Remove this once Process is implemented
+        io << Process.pid
+        io << '-'
+      {% end %}
+
+      io << Random.rand(0x100000000).to_s(36)
+
+      io << suffix
+    end
+
+    File.join(dir, name)
+  end
+
+  # Returns a fully-qualified path to a temporary file.
+  # The optional *suffix* is appended to the file name.
+  #
+  # ```
+  # File.tempname # => "/tmp/20171206-1234-449386"
+  # File.tempname(".sock") # => "/tmp/20171206-1234-449386.sock"
+  # ```
+  def self.tempname(suffix : String? = nil, *, dir : String = Dir.tempdir)
+    tempname(prefix: nil, suffix: suffix, dir: dir)
+  end
+
+  # Creates a temporary file.
+  #
+  # ```
+  # tempfile = File.tempfile("foo", ".bar")
+  # tempfile.delete
+  # ```
+  #
+  # *prefix* and *suffix* are appended to the front and end of the file name, respectively.
+  # These values may contain directory separators.
+  #
+  # The file will be placed in *dir* which defaults to the standard temporary directory `Dir.tempdir`.
+  #
+  # *encoding* and *invalid* are passed to `IO#set_encoding`.
+  #
+  # It is the caller's responsibility to remove the file when no longer needed.
+  def self.tempfile(prefix : String?, suffix : String?, *, dir : String = Dir.tempdir, encoding = nil, invalid = nil)
+    fileno, path = Crystal::System::File.mktemp(prefix, suffix, dir)
+    new(path, fileno, blocking: true, encoding: encoding, invalid: invalid)
```

Comment:
Why blocking true?

Reply:
Not sure... I just copied it form `Tempfile#initialize`.

Reply:
all files use blocking IO anyway because `O_NONBLOCK` has no effect on files.